### PR TITLE
[SYCL][Reduction] Fix race condition in withAuxHandler

### PR DESCRIPTION
**Pull request by aelovikov-intel**

### CODE REVIEW

File: sycl/include/sycl/reduction.hpp [hunk 1]
```diff
@@ -821,8 +821,9 @@ auto make_reduction(RedOutVar RedVar, RestTy &&...Rest) {
 namespace reduction {
 inline void finalizeHandler(handler &CGH) { CGH.finalize(); }
 template <class FunctorTy> void withAuxHandler(handler &CGH, FunctorTy Func) {
-  CGH.finalize();
+  auto E = CGH.finalize();
```
Comment (steffenlarsen) [inline sycl/include/sycl/reduction.hpp:824]: `finalize` returns an `event`, right? If so I would prefer to just have the type explicit here.

**Review by steffenlarsen — APPROVED** [review]
Minor comment, but otherwise a good catch! 😄

File: sycl/include/sycl/reduction.hpp (hunk 1 above)
Comment (aelovikov-intel) [inline sycl/include/sycl/reduction.hpp:824]:
```suggestion
  event E = CGH.finalize();
```

Comment (aelovikov-intel) [reply sycl/include/sycl/reduction.hpp]: Done.

**Review by steffenlarsen — APPROVED** [review]

**Comment by aelovikov-intel** [comment]
Failures:
  SYCL :: Basic/barrier_order.cpp - https://github.com/intel/llvm/issues/7330

@steffenlarsen is addressing those three with his PRs:
  SYCL :: BFloat16/bfloat16_builtins.cpp
  SYCL :: Matrix/element_wise_wi_marray_legacy.cpp
  SYCL :: Matrix/joint_matrix_tensorcores_legacy.cpp

**Pull request merged**